File: codeforces/1500~1599/1584/d.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

i64 query(int l, int r) {
  cout << "? " << l << " " << r << endl;
  i64 x = 0;
  cin >> x;
  return x;
}

int find_last(int l, int r, int x) {
  int ans = l;
  while (l <= r) {
    int mid = l + (r-l) / 2;
    i64 y = query(1, mid);
    if (y <= x) {
      l = mid+1;
      ans = mid;
    } else {
      r = mid-1;
    }
  }
  return ans;
}

void answer(int i, int j, int k) {
  cout << "! " << i << " " << j << " " << k << endl;
}

void solve() {
  int n = 0;
  cin >> n;

  i64 max_val = query(1, n);
  int i = find_last(1, n, 0);
  i64 len1 = max_val - query(i+1, n);
  int j = i + len1 + 1;
  i64 first_part = (len1 + 1) * len1 / 2;
  i64 second_part = max_val - first_part;
  // x * x + x = q * 2
  // x = (-1 + sqrt(4 * q * 2 + 1)) / 2
  int len2 = (sqrt(8 * second_part + 1) - 1) / 2;
  int k = j + len2;
  answer(i, j, k);
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  int T = 0;
  cin >> T;
  for (int t = 0; t < T; ++t) {
    solve();
  }
  return 0;
}

File: codeforces/1500~1599/1584/d.rs
use std::io::{stdin, stdout, Write};
use std::str::FromStr;
use std::fmt::Debug;

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn query(l: i64, r: i64) -> i64 {
    println!("? {} {}", l, r);
    stdout().flush().unwrap();
    read_int()
}

fn find_i(l: i64, r: i64) -> i64 {
    let mut l = l;
    let mut r = r;
    let mut ans = l;
    while l <= r{
        let mid = l + (r-l) / 2;
        let result = query(1, mid);
        if result <= 0 {
            l = mid + 1;
            ans = mid;
        } else {
            r = mid - 1;
        }
    }
    ans
}

fn solve() {
    let n = read_int::<i64>();
    let i = find_i(1, n);
    let total_inversion = query(1, n);
    let len1 = total_inversion - query(i+1, n);
    let j = i + len1 + 1;
    let p1 = len1 * (len1 + 1) / 2;
    let p2 = total_inversion - p1;
    let len2 = (((8 * p2 + 1) as f64).sqrt() as i64 - 1) / 2;
    let k = j + len2;
    println!("! {} {} {}", i, j, k);
    stdout().flush().unwrap();
}

fn main() {
    let n = read_int::<i32>();
    for _t in 0..n {
        solve();
    }
}
